RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHGCalIDVarProducer.cc

```diff
@@ -0,0 +1,139 @@
+
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/RecoCandidate/interface/RecoEcalCandidateIsolation.h"
+#include "DataFormats/EgammaReco/interface/SuperCluster.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/CaloRecHit/interface/CaloCluster.h"
+#include "DataFormats/CaloRecHit/interface/CaloClusterFwd.h"
+
+#include "DataFormats/EgammaReco/interface/SuperCluster.h"
+#include "DataFormats/EgammaReco/interface/SuperClusterFwd.h"
+
+#include "DataFormats/RecoCandidate/interface/RecoEcalCandidate.h"
+#include "DataFormats/RecoCandidate/interface/RecoEcalCandidateFwd.h"
+
+#include "DataFormats/ParticleFlowReco/interface/PFRecHit.h"
+
+#include "RecoEgamma/EgammaTools/interface/HGCalShowerShapeHelper.h"
+#include "RecoEgamma/EgammaTools/interface/HGCalClusterTools.h"
+
+class EgammaHLTHGCalIDVarProducer : public edm::stream::EDProducer<> {
+public:
+  explicit EgammaHLTHGCalIDVarProducer(const edm::ParameterSet&);
+  ~EgammaHLTHGCalIDVarProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  class PCAAssocMap {
```

Comment:
The way this class is used below should be fine - but the interfaces it presents exposes makes it a bit error prone.

Returning a `std::unique_ptr<...>&` has (at least ?) two problems:
  - whether ownership is taken from the object depends on what the reference is bound to (a reference, a value, nothing);
  - if bound to a reference, I think the behaviour relies on the object itself still being alive when the `unique_ptr&` is used; if it gets destroyed before the usage, one gets a dangling pointer
  - since one could end up with multiple references to the same unique_ptr, any could change or destroy it wihout the others knowing it

If the intended behaviour is that of shared ownership, it would be better to hold and return a `std::sharet_ptr`.
If the intended behaviour is to transfer ownership in a well-defined place (which I think is the case here), it would be better if
  - `initMap()` returns `void` (the return value is not used below)
  - `operator()` is renames to something like `getMap()` and returns `std::unique_ptr<...>` by value, not by reference

Reply:
you're completely correct of course, will adjust